File: packages/vscode/src/cli-locator.ts
import { CliNotFound, grokHome } from "@groks-beard/core"
import { Effect } from "effect"

export type LocateGrokInput = {
  readonly cliPath?: string
  readonly env: Record<string, string | undefined>
  readonly exists: (path: string) => boolean
  readonly win?: boolean
}

const join = (dir: string, name: string): string => `${dir.replace(/[\\/]+$/, "")}/${name}`

export const grokBinaryName = (win: boolean): string => (win ? "grok.exe" : "grok")

export const resolveSpawnTarget = (
  candidate: string,
  exists: (path: string) => boolean,
): string => {
  if (/\.(cmd|bat)$/i.test(candidate)) {
    const exe = candidate.replace(/\.(cmd|bat)$/i, ".exe")
    if (exists(exe)) return exe
  }
  return candidate
}

export const candidateGrokPaths = (input: LocateGrokInput): ReadonlyArray<string> => {
  const win = input.win ?? false
  const bin = grokBinaryName(win)
  const out: Array<string> = []
  if (input.cliPath !== undefined && input.cliPath !== "") out.push(input.cliPath)
  out.push(join(join(grokHome(input.env), "bin"), bin))
  const pathEnv = input.env.PATH ?? input.env.Path ?? ""
  const sep = win ? ";" : ":"
  for (const dir of pathEnv.split(sep)) {
    if (dir.trim() === "") continue
    out.push(join(dir, bin))
    if (win) out.push(join(dir, "grok.cmd"))
  }
  return out
}

export const locateGrokCli = (input: LocateGrokInput): Effect.Effect<string, CliNotFound> => {
  const searched = candidateGrokPaths(input)
  const hit = searched.find((path) => input.exists(path))
  if (hit === undefined) return Effect.fail(new CliNotFound({ searched: [...searched] }))
  return Effect.succeed(resolveSpawnTarget(hit, input.exists))
}

File: core/src/main/scala/groksbeard/core/CliLocator.scala
package groksbeard.core

final case class LocateGrok(
    cliPath: Option[String],
    env: String => Option[String],
    win: Boolean,
    exists: String => Boolean,
)

object CliLocator:
  def binaryName(win: Boolean): String = if win then "grok.exe" else "grok"

  def join(dir: String, name: String): String =
    s"${dir.replaceAll("[\\\\/]+$", "")}/$name"

  def resolveSpawnTarget(candidate: String, exists: String => Boolean): String =
    val lower = candidate.toLowerCase
    if lower.endsWith(".cmd") || lower.endsWith(".bat") then
      val exe = candidate.substring(0, candidate.length - 4) + ".exe"
      if exists(exe) then exe else candidate
    else candidate

  def candidates(input: LocateGrok): List[String] =
    val bin = binaryName(input.win)
    val out = List.newBuilder[String]
    input.cliPath.filter(_.nonEmpty).foreach(out += _)
    out += join(join(GrokHome(input.env), "bin"), bin)
    val pathEnv = input.env("PATH").orElse(input.env("Path")).getOrElse("")
    val sep     = if input.win then ";" else ":"
    pathEnv.split(sep).foreach { dir =>
      if dir.trim.nonEmpty then
        out += join(dir, bin)
        if input.win then out += join(dir, "grok.cmd")
    }
    out.result()
  end candidates

  def locate(input: LocateGrok): Either[List[String], String] =
    val searched = candidates(input)
    searched.find(input.exists) match
      case None      => Left(searched)
      case Some(hit) => Right(resolveSpawnTarget(hit, input.exists))
end CliLocator
